syncer/syncer_test.go

```diff
@@ -613,55 +716,62 @@ func (s *testSyncerSuite) TestColumnMapping(c *C) {
+	events = append(events, mockBinlogEvent{typ: DBCreate, args: []interface{}{"stest_3"}})
+	events = append(events, mockBinlogEvent{typ: TableCreate, args: []interface{}{"stest_3", "create table stest_3.log(id varchar(45))"}})
+	events = append(events, mockBinlogEvent{typ: TableCreate, args: []interface{}{"stest_3", "create table stest_3.t_2(name varchar(45), id bigint)"}})
+	events = append(events, mockBinlogEvent{typ: TableCreate, args: []interface{}{"stest_3", "create table stest_3.a(id int)"}})
+
+	createEvents := s.generateEvents(events, c)
 
-	dmls := []struct {
-		sql    string
+	// dmls
+	type dml struct {
+		events []*replication.BinlogEvent
 		column []string
 		data   []interface{}
-	}{
-		{"insert into stest_3.t_2(name, id) values (\"ian\", 10)", []string{"name", "id"}, []interface{}{"ian", int64(1<<59 | 3<<52 | 2<<44 | 10)}},
-		{"insert into stest_3.log(id) values (\"10\")", []string{"id"}, []interface{}{"test:10"}},
-		{"insert into stest_3.a(id) values (10)", []string{"id"}, []interface{}{int32(10)}},
 	}
 
-	dropTableSQLs := []string{
-		"drop table stest_3.log,stest_3.t_2,stest_3.a",
-		"drop database stest_3",
-	}
+	dmls := make([]dml, 0, 3)
 
-	for _, sql := range createTableSQLs {
-		s.db.Exec(sql)
-	}
+	evs := s.generateEvents([]mockBinlogEvent{{typ: Write, args: []interface{}{uint64(8), "stest_3", "t_2", []byte{mysql.MYSQL_TYPE_STRING, mysql.MYSQL_TYPE_LONG}, [][]interface{}{{"ian", int32(10)}}}}}, c)
+	dmls = append(dmls, dml{events: evs, column: []string{"name", "id"}, data: []interface{}{"ian", int64(1<<59 | 3<<52 | 2<<44 | 10)}})
 
-	for i := range dmls {
-		s.db.Exec(dmls[i].sql)
-	}
+	evs = s.generateEvents([]mockBinlogEvent{{typ: Write, args: []interface{}{uint64(9), "stest_3", "log", []byte{mysql.MYSQL_TYPE_STRING}, [][]interface{}{{"10"}}}}}, c)
+	dmls = append(dmls, dml{events: evs, column: []string{"id"}, data: []interface{}{"test:10"}})
 
-	for _, sql := range dropTableSQLs {
-		s.db.Exec(sql)
+	evs = s.generateEvents([]mockBinlogEvent{{typ: Write, args: []interface{}{uint64(10), "stest_3", "a", []byte{mysql.MYSQL_TYPE_LONG}, [][]interface{}{{int32(10)}}}}}, c)
+	dmls = append(dmls, dml{events: evs, column: []string{"id"}, data: []interface{}{int32(10)}})
+
+	dmlEvents := make([]*replication.BinlogEvent, 0, 15)
+	for _, dml := range dmls {
+		dmlEvents = append(dmlEvents, dml.events...)
 	}
 
-	p, err := utils.GetParser(s.db, false)
+	// drop tables and db
+	events = events[:0]
+	events = append(events, mockBinlogEvent{typ: TableDrop, args: []interface{}{"stest_3", "log"}})
+	events = append(events, mockBinlogEvent{typ: TableDrop, args: []interface{}{"stest_3", "t_2"}})
+	events = append(events, mockBinlogEvent{typ: TableDrop, args: []interface{}{"stest_3", "a"}})
+	events = append(events, mockBinlogEvent{typ: DBDrop, args: []interface{}{"stest_3"}})
+	dropEvents := s.generateEvents(events, c)
+
+	db, mock, err := sqlmock.New()
+	mock.ExpectQuery("SHOW GLOBAL VARIABLES LIKE").
```

Comment:
seems we can add a function to generate mock result for sql_mode

Reply:
Done